### ISSUE TITLE
"`ResourceSupervisor`"

### ISSUE THREAD

**Issue by armanbilge** [typelevel/cats-effect#3376]
This is an idea @TimWSpence and I cooked up once on Discord, starting from https://github.com/typelevel/fs2/pull/2300#issuecomment-1325010963.

The idea is very similar to the existing `Supervisor` which allows to `start` a fiber, and bind its lifecycle to some outer scope. Here, we want to `allocate` a `Resource` and bind its lifecycle to the outer scope.

```scala
ResourceSupervisor[F].use { supervisor =>
  resourceA.allocateSupervised(supervisor).flatMap { (a, release) =>
    // do stuff
  }
}
```

`allocateSupervised` (or `superviseAllocated` ?) would work like `allocated`, except in case the `release` is not invoked for some reason, it will be cleaned up once the supervisor scope ends.

tl;dr `start` is to `allocated` what `supervise` is to (we need something here :)

**Comment by armanbilge** [comment]
Some overlap with:
- https://github.com/typelevel/cats-effect/issues/1345

**Comment by Jasper-M** [comment]
I made something very much like this for a private project. But I actually only needed `useSupervised` instead of `allocateSupervised`, i.e. you can't release prematurely.
